### PR TITLE
Implement `scroll_view` widget

### PR DESCRIPTION
**Pull request by nicoburns**
~Builds on top of https://github.com/linebender/xilem/pull/140 which should be merged first.~ I've now made this PR independent of #140

## Notes

- I'm not very confident about the approach of changing the origin or clipping the painting.
- I had to change the `set_origin` method take `&mut WidgetState` rather than `&mut LayoutCx`. Because otherwise I couldn't call it from `event` and had to request layout every time the scroll position changed.

## Video

https://github.com/linebender/xilem/assets/1007307/dfc921fc-daf7-4010-87e6-81fe3a2e7d7f

### CODE REVIEW

File: src/view/scroll_view.rs [hunk 1]
```diff
@@ -58,27 +62,22 @@ where
...
+                .rebuild(cx, &prev.child, &mut state.0, &mut state.1, child_element)
         })
     }
 
-    fn event(
+    fn message(
         &self,
         id_path: &[Id],
         state: &mut Self::State,
         event: Box<dyn Any>,
```
Comment (Philipp-M) [inline src/view/scroll_view.rs:77]:
```suggestion
        message: Box<dyn Any>,
```

File: src/view/scroll_view.rs [hunk 2]
```diff
@@ -58,27 +62,22 @@ where
...
         app_state: &mut T,
-    ) -> EventResult<A> {
+    ) -> MessageResult<A> {
         let tl = &id_path[1..];
```
Comment (Philipp-M) [inline src/view/scroll_view.rs:80]:
```suggestion
```

File: src/view/scroll_view.rs [hunk 3]
```diff
@@ -58,27 +62,22 @@ where
...
-        self.child.event(tl, &mut state.1, event, app_state)
+        self.child.message(tl, &mut state.1, event, app_state)
```
Comment (Philipp-M) [inline src/view/scroll_view.rs:81]: I know this is existing code, but I guess it makes sense to make this safer at this point.
(not sure how I can do a multiline suggestion across additions and existing code)

```suggestion
        match id_path {
            [child_id, rest_path @ ..] if *child_id == state.0 => {
                self.child
                    .message(rest_path, &mut state.1, message, app_state)
            }
            _ => MessageResult::Stale(message),
        }
```

File: src/widget/core.rs [hunk 4]
```diff
@@ -94,7 +95,7 @@ pub struct Pod {
 }
 
 #[derive(Debug)]
-pub(crate) struct WidgetState {
+pub struct WidgetState {
```
Comment (Philipp-M) [inline src/widget/core.rs:98]: Out of curiosity, why this change?

File: src/widget/core.rs [hunk 5]
```diff
@@ -467,18 +486,23 @@ impl Pod {
...
-            cx.request_paint();
+            parent_state.flags |= PodFlags::REQUEST_PAINT;
+            parent_state.flags |= PodFlags::VIEW_CONTEXT_CHANGED;
 
             self.state.flags.insert(PodFlags::VIEW_CONTEXT_CHANGED);
         }
     }
 
+    /// Get the widgets size (as returned by the layout method)
+    pub fn get_size(&mut self) -> Size {
```
Comment (Philipp-M) [inline src/widget/core.rs:502]: I think we should omit `get_` prefixes to stay consistent 
```suggestion
    pub fn size(&mut self) -> Size {
```

File: src/widget/scroll_view.rs [hunk 6]
```diff
@@ -47,41 +44,26 @@ impl ScrollView {
...
-        self.child.event(cx, &child_event);
+    fn event(&mut self, cx: &mut EventCx, event: &Event) {
+        // Pass event through to child
+        self.child.event(cx, event);
+
+        // Handle scroll wheel events
         if !cx.is_handled() {
-            if let RawEvent::MouseWheel(mouse) = event {
-                let new_offset = (self.offset + mouse.wheel_delta.y).max(0.0);
+            if let Event::MouseWheel(mouse) = event {
```
Comment (Philipp-M) [inline src/widget/scroll_view.rs:55]: Out of a UX perspective it would be interesting (not entirely sure how this is done in web, but I think it's similar as here?), how to handle nested scroll-views.

When I'm getting this right, the nested scroll view handles the scroll event, unless it doesn't scroll (e.g. is at the end).

Just thinking out loud, not suggesting anything.

**Review by Philipp-M — COMMENTED** [review]
Thanks :)

I think it generally makes sense to add small examples showing the widgets in action (I personally like self-contained examples showing each widgets capabilities on its own, for this maybe a lorem ipsum text in a scroll view), but not a requirement.

I'm not (yet) deep in layouting and vello, so I may be wrong here: Wouldn't it make sense, instead of setting the child origin, to use the vello layer with a transform that has the offset instead?

File: src/view/scroll_view.rs (hunk 1 above)
Comment (nicoburns) [reply src/view/scroll_view.rs:77]: Could you explain the difference between "message" and "event" to me? I'm very surprised that there are two functions with identical signatures that seemingly do a similar thing!

File: src/widget/core.rs (hunk 4 above)
Comment (nicoburns) [reply src/widget/core.rs:98]: This is to allow `set_origin` to be called with multiple different context types. As `set_origin` is public, and now takes `WidgetState` as a parameter, `WidgetState` also needs to be public.

File: src/widget/scroll_view.rs (hunk 6 above)
Comment (nicoburns) [reply src/widget/scroll_view.rs:55]: I believe that ideally it would be a little more sophisticated than that, and that the inner view could handle part of a scroll event (some of the pixel delta) and leave the rest for an outer scroll view (perhaps by making the event mutable?). But you suggestion would probably cover the basics.

**Comment by nicoburns** [comment]
> I'm not (yet) deep in layouting and vello, so I may be wrong here: Wouldn't it make sense, instead of setting the child origin, to use the vello layer with a transform that has the offset instead?

I believe that would draw the widget in the correct place, but would not allow input to work correctly. Whereas set_origin applies to both.

File: src/view/scroll_view.rs (hunk 1 above)
Comment (Philipp-M) [reply src/view/scroll_view.rs:77]: I think originally (in the druid [xilem_tokio](https://github.com/linebender/druid/tree/xilem_tokio) branch) it was called event (and was put unmodified into this repo), and was renamed later to message, I'm not entirely sure (Raph probably has a better answer), just a naming thing.

I think it's because `message` kind of implies that it "messages" a general payload, while event doesn't necessarily.

**Comment by Philipp-M** [comment]
> I believe that would draw the widget in the correct place, but would not allow input to work correctly. Whereas set_origin applies to both.

That's a good point, might be interesting to research something like virtualized pointer events or similar that can be modified while traversing down the tree (in `Pod::event`).
Hah just saw in that moment, that was done previously like that^^, what speaks against this approach for now?

**Review by derekdreery — APPROVED** [review]
Awesome, thanks.

**Comment by DJMcNab** [comment]
@nicoburns are you expecting to revive this branch? It seems to be on the Xilem repo and not deleted.

**Comment by nicoburns** [comment]
No. #155 was merged instead

**Comment by DJMcNab** [comment]
I'll delete the branch then
